Add fluent configuration of ffoptions per run

BobSilent opened the pull request:
Thanks for closing #272 so quick,! I would like to ask for an additional configuration extension here:

Add Configure on FFMpegArgumentProcessor to fuently configure ffoptions per run.

Currently the GlobalFFOptions allows an action delegate with the configure option
```csharp
GlobalFFOptions.Configure(options => options.BinaryFolder = "./bin");
``` 

What I am missing, is the same for the per run option. I can pass a new ffoption to override all existing, but i cannot extend reconfigure the existing options per run.
This PR enables the following:
1. setting default global options
2. updating the settings on run base

```csharp
    GlobalFFOptions.Configure(options => options.BinaryFolder = "GlobalBinaryFolder...");
    GlobalFFOptions.Configure(options => options.WorkingDirectory = "GlobalWorkingDirectory");

    FFMpegArguments
        .FromFileInput(inputPath)
        .OutputToFile(outputPath, false, options => options.WithFastStart())
        .Configure(options => options.WorkingDirectory = "CurrentRun WorkingDirectory")
        .Configure(options => options.TemporaryFilesFolder = "CurrentRun TempFolder")
        .ProcessSynchronously();
```

if ok, i would extend the documentation as well

BobSilent committed:
Add Configure on FFMpegArgumentProcessor to fuently configure ffoptions per run.

rosenbjerg commented:
Hi @BobSilent 
Could you add a test that verifies that the behaviour is correct? though it does indeed look correct.
That way, we'll also protect against the feature being broken by accident later on :)

And sorry for the delay. I have been quite busy at work

BobSilent committed:
Update Docu

BobSilent committed:
Added Tests

BobSilent commented:
I updated the documentation.

For testing I added the FluentAssertions Package, I hope this is okay, otherwise I will remove it and update PR.

From test perspective I am not totally happy with the test.
Ideally I could Inject an Instance Mock and see that the Instance gets configured correctly. But doing this would need more changes, ideally adding an interface to Instance.
This could also improve some other test, to only check ffmpeg call without the need to explicitly execute it.

rosenbjerg commented:
Hi again @BobSilent, the tests and  look great :)
I can definitely follow you in your argument for mocking Instance. I'll definitely think about adding this to the Instances library

Thank you for the contribution, it'll be included in the next release of the library

rosenbjerg commented:
One problem with the tests is that since the GlobalConfiguration is modified, all tests run that require the default configuration will fail if they run after the new tests that modify GlobalConfiguration. So we'll need to find another way to test this, which doesn't change the GlobalConfiguration to something invalid. Maybe we don't even need to change it? We can just verify that it still has the same value when passing `null` to the action. That way we perform the same checks, so the testing should be equal, but we avoid modifying the default values of GlobalConfiguration, which is needed for the other tests :)

rosenbjerg commented on FFMpegCore.Test/FFMpegArgumentProcessorTest.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+            var globalConfig = "Whatever";
+            GlobalFFOptions.Configure(new FFOptions { WorkingDirectory = globalConfig, TemporaryFilesFolder = globalConfig, BinaryFolder = globalConfig });
+
```
Running this line will break other tests that are getting executed after this one

BobSilent commented:
ok, you are right, let me think about it i will come up with a proposal

rosenbjerg commented on FFMpegCore.Test/FFMpegArgumentProcessorTest.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+
+            var globalWorkingDir = "Whatever";
+            GlobalFFOptions.Configure(new FFOptions { WorkingDirectory = globalWorkingDir });
```
Running this line will break other tests that are getting executed after this one

rosenbjerg commented on FFMpegCore.Test/FFMpegArgumentProcessorTest.cs:
```diff
@@ -0,0 +1,84 @@
+                .OutputToFile("");
+
+            var globalWorkingDir = "Whatever";
+            GlobalFFOptions.Configure(new FFOptions { WorkingDirectory = globalWorkingDir });
```
Running this line will break other tests that are getting executed after this one

rosenbjerg reviewed (changes requested):
We'll need to change the lines I've commented on, but otherwise LGTM :)

BobSilent committed:
reset GlobalFFOptions to null after testing

BobSilent commented:
@rosenbjerg is this okay, to keep the test but reset the field after testing?
https://github.com/rosenbjerg/FFMpegCore/blob/8cacf074fd830032279df9f05d88b0472e5ac011/FFMpegCore.Test/FFMpegArgumentProcessorTest.cs#L10-L16

rosenbjerg commented:
Looks good. Thank you for the contríbution!

rosenbjerg commented:
v.4.7.0 containing this feature has been released, Thanks again for the contribution
